### ISSUE TITLE
pyfilesystem2 (fs) is no longer maintained; find a replacement or avoid dependency?

### ISSUE THREAD

**Issue by anthrotype** [fonttools/fonttools#3620]
the `fs` module that fontTools.ufoLib uses to implement .ufoz (UFO inside a .zip file) hasn't received an update in more than 2 years, it's basically unmaintained, see: https://github.com/PyFilesystem/pyfilesystem2/issues/571

There is a new module called https://github.com/fsspec/filesystem_spec which may be a suitable replacement and is more actively maintained. Maybe we should try to use that.
It does have a ZipFileSystem implementation:
https://github.com/fsspec/filesystem_spec/blob/master/fsspec/implementations/zip.py

Or alternatively we could reimplement our own filesystem-like wrapper around a zip file.

I'd prefer option 1), or whichever is easier.

**Comment by justvanrossum** [comment]
I just had a brief look at fsspec: it looks very good. Also in favor of option 1).

**Comment by a-detiste** [comment]
Hi, keeping `fs` alive is a liability for the distribution. In Debian I see that `fonttools` will soon be the very last user of `fs`. Replacing it's usage with a maintained library is hugely appreciated.

**Comment by behdad** [comment]
I remember when I prototyped using the `fs` module first in `ufoLib`, I also implemented a dummy fallback. Maybe we can find that and use it if `fs` is not available for now, until port to `fsspec` happens.

**Comment by behdad** [comment]
> I remember when I prototyped using the `fs` module first in `ufoLib`, I also implemented a dummy fallback. Maybe we can find that and use it if `fs` is not available for now, until port to `fsspec` happens.

Looks like I was wrong: https://github.com/behdad/ufoLib/commit/acfe65b312c3c7fe5e31b2040b406f145e396ebb

**Comment by skef** [comment]
@behdad looks like Python 3.13 is raising inconvenient (for AFDKO) warnings from fs now:

```
../../py/lib/python3.13/site-packages/fs/__init__.py:4
  /home/skef/py/lib/python3.13/site-packages/fs/__init__.py:4: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools<81.
    __import__("pkg_resources").declare_namespace(__name__)  # type: ignore

../../py/lib/python3.13/site-packages/fs/__init__.py:4
  /home/skef/py/lib/python3.13/site-packages/fs/__init__.py:4: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('fs')`.
  Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
    __import__("pkg_resources").declare_namespace(__name__)  # type: ignore

../../py/lib/python3.13/site-packages/fs/opener/__init__.py:6
  /home/skef/py/lib/python3.13/site-packages/fs/opener/__init__.py:6: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('fs.opener')`.
  Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
    __import__("pkg_resources").declare_namespace(__name__)  # type: ignore

../../py/lib/python3.13/site-packages/pkg_resources/__init__.py:2558
  /home/skef/py/lib/python3.13/site-packages/pkg_resources/__init__.py:2558: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('fs')`.
  Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
    declare_namespace(parent)
```
We have a couple of janky tests that compare console output to a reference that will be tricky to work around. Not your folks problem but it would be good to know what the plan is for fs medium term, as our users will likely start seeing these.

**Comment by behdad** [comment]
I'll try to have a look in a week and see if I can kill it or port it.

**Comment by skef** [comment]
> We have a couple of janky tests that compare console output to a reference that will be tricky to work around

Correction: Quite easy to work around. Just learned about `PYTHONWARNINGS=ignore` .

**Comment by anthrotype** [comment]
I started looking at this last week. fsspec supports all the features of pyfilesystem2 and is well maintained. The API is slightly different (some methods do the same thing but are named differently, or e.g. the local filesystem always uses absolute paths whereas pyfs2 would let one create a filesystem instance from a specific directory and use relative paths), but we can easily adapt our code to it, perhaps define a wrapper that exposes the same interface (or the subset that we currently use) as pyfilesystem2 which internally uses fsspec object so we don't need to update all the call sites. I can work on that, unless @behdad beats me to it

**Comment by behdad** [comment]
@anthrotype Please go ahead. I have a few other things on my plate.

**Comment by musicinmybrain** [comment]
I’ll add my name to the list of people who are excited you are working on this. I am working through making [`python-fs`](https://src.fedoraproject.org/rpms/python-fs) build and pass (enough) tests with Python 3.14 in Fedora Rawhide, and it’s hard to work on an upstream that still hasn’t merged PR’s for changes in Python 3.12.

**Comment by a-detiste** [comment]
>  in Fedora Rawhide

In Debian we are in freeze period, but yeah this _will_ be hugely appreciated.

Status of  `python-fs` came in light because of the `six` removal crusade ... https://wiki.debian.org/Python3-six-removal

I don't know/think Fedora is tackling this yet. Another worrysome key library is `dateutil`. Otherwise it's going smoothly slow & steady now.

**Comment by musicinmybrain** [comment]
> In Debian I see that `fonttools` will soon be the very last user of `fs`.

This is already true in Fedora, by the way.

**Comment by a-detiste** [comment]
we have this very outdated `organize`thing
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1082579

Le sam. 14 juin 2025 à 15:07, Ben Beasley ***@***.***> a
écrit :

**Comment by anthrotype** [comment]
@madig wrote https://github.com/fonttools/fonttools/pull/3871#issuecomment-3001039127

> There is another solution: Drop UFOZ support and fs/fsspec along with it. It seems to be a very niche feature.

Well no, I think we should keep UFOZ support, even though it's probably not very much used nowadays that people are more familiar with git. It's in the UFO spec and it has its legitimate use cases.

But we can try to do that without adding a new dependency on fsspec. It shouldn't be too hard to implement a fs-like wrapper around stdlib's `zipfile` module for reading single files lazily off a zip (without needing to extract all to a local directory). We already stage the writing of a UFOZ to a temporary directory and compress only upon closing the UFOWriter so not much would change there.

I feel a bit responsible for insisting on adding pyfilesytem2 to the list of fonttools dependencies seven years ago, so I think I should be the one that takes care of getting rid of it :)

Realistically, the ability to inject abitrary fs-compatible filesystems never materialized so I think nobody would probably notice that this stops working.

**Comment by madig** [comment]
Right. From a skim of the zipfile docs, it seems to be able to read and write files in place? How would you architect this? Reading would extract files on demand without extraction, writing would use a temporary directory and zip up at the end?

**Comment by anthrotype** [comment]
> Reading would extract files on demand without extraction, writing would use a temporary directory and zip up at the end?

pretty much, yes.

**Comment by madig** [comment]
Could you write directly into a zip? I think UFOWriter uses a temporary directory because you might need to write to the zip you read from, but you can't open it for writing without truncating it first? But then, you could write to a temporary zip and move it over at the end? Maybe? Maybe not on Windows?

**Comment by anthrotype** [comment]
It can but only works if you are creating a new zip, not if updating in place (maybe it can append files but we sometimes also need to remove files so that won't work). I think it's just best to write to a temporary location whether or not we are updating in-place or writing a new UFOZ, and only zip at the end and move when everything is successful, easier and safer.

**Comment by anthrotype** [comment]
Fixed by #3885 

from now on, installing fonttools with the `[ufo]` extra will no longer automatically install `fs`; a subset of the latter's api was reimplemented in `fontTools.misc.filesystem` using stdlib-only modules (pathlib, zipfile, shutil, etc.). ufoLib will still continue to work with the upstream `fs` for a while (and we even test with/without `fs`) and clients who wish to continue using `fs` alongside fonttools can do so but will have to depend on `fs` directly instead of via `fonttools[ufo]`.
